src/hotspot/share/runtime/handshake.cpp

```diff
@@ -786,6 +792,17 @@ bool HandshakeState::suspend() {
   } else {
     SuspendThreadHandshake st;
     Handshake::execute(&st, _handshakee);
+
+    if (st.did_suspend() && _handshakee->is_virtual()) {
+      // Thread suspension works under JvmtiVTMSTransitionDisabler protection, so we need to wait
+      // for virtual thread to reach a safe state before the JVMTI SuspendThread is returned.
+      while (_handshakee->thread_state() != _thread_blocked &&
+             _handshakee->thread_state() != _thread_in_native) {
```

Comment:
We already wait for the target to reach a handshake-safe state when executing the `SuspendThreadHandshake`. It’s just that we don’t wait until the target suspends in `do_self_suspend()` since that can take arbitrarily long (e.g. target just waiting in native).